File: PERSISTENCE/Core Data/[Objective-C] Core Data/coredata_cars/coredata_cars/carTableViewController.m
//
//  carTableViewController.m
//  coredata_cars
//
//  Created by Luis Andres Castillo Hernandez on 11/18/15.
//  Copyright © 2015 devbros. All rights reserved.
//

#import "carTableViewController.h"
#import "addCarViewController.h"


#import <CoreData/CoreData.h>

@interface carTableViewController ()

@property (strong) NSMutableArray * devices;

@end

@implementation carTableViewController


//MARK:

-(NSManagedObjectContext *) managedObjectContext
{
    NSManagedObjectContext *context = nil;
    id delegate = [[UIApplication sharedApplication] delegate];
    
    if([delegate  performSelector:@selector(managedObjectContext)])
    {
        context = [delegate managedObjectContext];
    }
    
    return context;
    
}//eom


//MARK: View Loading

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
//     self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
     self.navigationItem.leftBarButtonItem = self.editButtonItem;
}

-(void)viewDidAppear:(BOOL)animated
{

    NSManagedObjectContext * managedObjectContext = [self managedObjectContext];
    NSFetchRequest *fetchRequest = [[NSFetchRequest alloc]initWithEntityName:@"Device"];
    
    //updating devices from core data information
    self.devices = [[managedObjectContext executeFetchRequest:fetchRequest error:nil]mutableCopy];
    
    //reloading table view data
    [self.tableView reloadData];
    
}//eom

//MARK: Memory

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {

    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.devices.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{

    static NSString *cellID = @"carCell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellID forIndexPath:indexPath];
    
    
    int currRow = (int)indexPath.row;
    
    //getting data
    NSManagedObjectModel * device = [self.devices objectAtIndex:currRow];
    
    //getting car info
    NSString *currCarMake   = [device valueForKey:@"carMake"];
    NSString *currCarModel  = [device valueForKey:@"carModel"];
    NSString *currCarYear   = [device valueForKey:@"carYear"];
    
    NSString *labelText = [NSString stringWithFormat:@"%@ %@", currCarMake, currCarModel];
    //updating title label
    [cell.textLabel setText:labelText];
    
    //updating detail label
    [cell.detailTextLabel setText:currCarYear];
    
    return cell;
}

//MARK: Edit

// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}


// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle
forRowAtIndexPath:(NSIndexPath *)indexPath {
    
    
    NSManagedObjectContext * context = [self managedObjectContext];
    
    //DELETE REQUEST
    if (editingStyle == UITableViewCellEditingStyleDelete)
    {
        
        //deleteing core data
        int index = (int)indexPath.row;
        id  itemID = [self.devices objectAtIndex:index];
        [context deleteObject:itemID];
        
        //checking if error saving values occurred
        NSError * error = nil;
        if( ![context save:&error])
        {
            NSLog(@"%@ %@", error, [error localizedDescription]);
        }
        
        //deleting from local list
        [self.devices removeObjectAtIndex:index];
        
        //deleteing from tableview
        [self.tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationFade];
        
        // Delete the row from the data source
//        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
        
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}

//MARK:

// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/


#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if([[segue identifier] isEqualToString:@"updateCar"])
    {
        //current index
//        NSInteger index = [[self.tableView indexPathForSelectedRow] row];
        
        NSManagedObjectModel * selectDevice = [self.devices objectAtIndex:[[self.tableView indexPathForSelectedRow] row]];
        addCarViewController *addView = segue.destinationViewController;
        addView.device = selectDevice;
        
    }
    else if([[segue identifier] isEqualToString:@"addCar" ])
    {
        
    }
    
    
}//eom


@end

File: PERSISTENCE/Core Data/[Swift] Core Data/coredata_cars/coredata_cars/carTableViewController.swift
//
//  carTableViewController.swift
//  coredata_cars
//
//  Created by Luis Andres Castillo Hernandez on 11/20/15.
//  Copyright © 2015 devbros. All rights reserved.
//

import UIKit
import CoreData

class carTableViewController: UITableViewController
{
    var List: Array<AnyObject> = []
    
      //MARK: view loading
    
    override func viewDidLoad()
    {
        super.viewDidLoad()

        // Uncomment the following line to preserve selection between presentations
        // self.clearsSelectionOnViewWillAppear = false

        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        // self.navigationItem.rightBarButtonItem = self.editButtonItem()
    }//eo-view

    override func viewDidAppear(_ animated: Bool)
    {
        
        let AppDel: AppDelegate = UIApplication.shared.delegate as! AppDelegate
        
        let Context: NSManagedObjectContext = AppDel.managedObjectContext
        let request = NSFetchRequest<NSFetchRequestResult>(entityName: "CarList")
        
        List = try! Context.fetch(request)
        tableView.reloadData()
    }//eo-view
    
    //MARK: memory
    override func didReceiveMemoryWarning()
    {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        return List.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
    {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "CarCell") as! carTableViewCell
        let data: NSManagedObject = List[indexPath.row] as! NSManagedObject
        
        cell.carMakeLabel?.text     = data.value(forKey: "carMake") as? String
        cell.carModelLabel?.text    = data.value(forKey: "carModel") as? String
        cell.carYearLabel?.text     = data.value(forKey: "carYear") as? String
        
        return cell
    }


    /*
    // Override to support conditional editing of the table view.
    override func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        // Return false if you do not want the specified item to be editable.
        return true
    }
    */

    
    // Override to support editing the table view.
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath)
    {

        let AppDel: AppDelegate = UIApplication.shared.delegate as! AppDelegate
        let Context: NSManagedObjectContext = AppDel.managedObjectContext

        if editingStyle == UITableViewCellEditingStyle.delete
        {

            Context.delete(List[indexPath.row] as! NSManagedObject)
            List.remove(at: indexPath.row)
            tableView.deleteRows(at: [indexPath], with: UITableViewRowAnimation.automatic)

        }

        var error: NSError? = nil

        do {

            try Context.save()

        }
        catch let error1 as NSError
        {
            error = error1
            print(error)
            abort()
        }



    }//eom


    /*
    // Override to support rearranging the table view.
    override func tableView(tableView: UITableView, moveRowAtIndexPath fromIndexPath: NSIndexPath, toIndexPath: NSIndexPath) {

    }
    */

    /*
    // Override to support conditional rearranging of the table view.
    override func tableView(tableView: UITableView, canMoveRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        // Return false if you do not want the item to be re-orderable.
        return true
    }
    */

    
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
        
        
        if segue.identifier == "editCar"
        {
            let selectedItem: NSManagedObject = List[self.tableView.indexPathForSelectedRow!.row] as! NSManagedObject
            
            let ViewCon: addCarViewController = segue.destination as! addCarViewController
            
            ViewCon.carmake     = selectedItem.value(forKey: "carmake") as! String
            ViewCon.carmodel    = selectedItem.value(forKey: "carmodel") as! String
            ViewCon.caryear     = selectedItem.value(forKey: "caryear") as! String
            
            ViewCon.existingItem = selectedItem
        }
        else if segue.identifier == "addCar"
        {
            
        }

    }//eo


}
